nucypher/utilities/networking.py

```diff
@@ -15,42 +15,141 @@
+    text content or None, suppressing all errors. Certificate is
+    needed if the remote URL source is self-signed.
     """
     try:
+        # 'None' or 'True' will verify self-signed certificates
+        response = requests.get(url, verify=certificate)
+    except RequestErrors:
+        return None
+    if response.status_code == 200:
+        return response.text
+
+
+def get_external_ip_from_default_teacher(network: str,
+                                         federated_only: bool = False,
+                                         log: Logger = IP_DETECTION_LOGGER,
+                                         registry: BaseContractRegistry = None
+                                         ) -> Union[str, None]:
+    if federated_only and registry:
+        raise ValueError('Federated mode must not be true if registry is provided.')
+    base_error = 'Cannot determine IP using default teacher'
+    try:
+        top_teacher_url = RestMiddleware.TEACHER_NODES[network][0]
+    except IndexError:
+        log.debug(f'{base_error}: No teacher available for network "{network}".')
+        return
+    except KeyError:
+        log.debug(f'{base_error}: Unknown network "{network}".')
+        return
+    if not registry:
+        # Registry is needed to perform on-chain staking verification.
+        registry = InMemoryContractRegistry.from_latest_publication(network=network)
+    teacher = Ursula.from_teacher_uri(teacher_uri=top_teacher_url,
+                                      registry=registry,
+                                      federated_only=federated_only,
+                                      min_stake=0)  # TODO: Handle customized min stake here.
+    client = NucypherMiddlewareClient()
+    try:
+        response = client.get(node_or_sprout=teacher, path=f"ping", timeout=2)  # TLS certificate logic within
+    except RestMiddleware.UnexpectedResponse:
+        # 404, 405, 500, All server response codes handled by will be caught here.
+        return  # Default teacher does not support this request - just move on.
+    if response.status_code == 200:
+        try:
+            ip = str(ip_address(response.text))
+        except ValueError:
+            error = f'Default teacher at {top_teacher_url} returned an invalid IP response; Got {response.text}'
+            raise UnknownIPAddress(error)
+        log.info(f'Fetched external IP address from default teacher ({top_teacher_url} reported {ip}).')
+        return ip
+
+
+def get_external_ip_from_known_nodes(known_nodes: FleetSensor,
+                                     sample_size: int = 3,
+                                     log: Logger = IP_DETECTION_LOGGER
+                                     ) -> Union[str, None]:
+    """
+    Randomly select a sample of peers to determine the external IP address
+    of this host. The first node to reply successfully will be used.
+    # TODO: Parallelize the requests and compare results.
```

Comment:
This would be cool - I'm pro this TODO since it provides consistency of IP address determination across the sampled nodes.

Reply:
Opened follow-up https://github.com/nucypher/nucypher/issues/2514